File: client/Network/OkHttp/app/src/main/java/com/example/okhttp/MainActivity.java
package com.example.okhttp;

import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.os.Bundle;
import android.os.Handler;
import android.util.JsonReader;
import android.util.JsonToken;
import android.util.Log;
import android.view.View;
import android.widget.Button;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;

import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.FormBody;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;

public class MainActivity extends AppCompatActivity implements InfoDialog.BottomSheetCallback {
    private static final String TAG = "MainActivity";
    private RecyclerView recyclerView;
    private RcAdapter adapter;
    private Button requestPost;
    private OkHttpClient client;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        client = new OkHttpClient();
        recyclerView = findViewById(R.id.json_list);
        requestPost = findViewById(R.id.request_post);



        LinearLayoutManager layoutManager = new LinearLayoutManager(getApplicationContext());
        recyclerView.setLayoutManager(layoutManager);
        adapter = new RcAdapter();
        recyclerView.setAdapter(adapter);

        requestPost.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                InfoDialog dialog = new InfoDialog(MainActivity.this);
                dialog.setBottomSheetCallback(MainActivity.this);
                dialog.show();

            }
        });

        FormBody formBody = new FormBody.Builder().build();

        Request request = new Request.Builder()
                .url("http://ec2-3-36-106-46.ap-northeast-2.compute.amazonaws.com:3001/movies")
                .get().build();

        client.newCall(request).enqueue(httpCallback);
    }

    private final Callback httpCallback = new Callback() {
        @Override
        public void onFailure(Call call, IOException e) {

        }

        @Override
        public void onResponse(Call call, Response response) throws IOException {
            if (response.body() != null) {
                String myJsonData = response.body().string();

                Log.e("fd", myJsonData);
                try {
                    JSONObject jsonObject = new JSONObject(myJsonData);
                    JSONArray jsonList = jsonObject.getJSONArray("data");
                    for (int i = 0; i < jsonList.length(); i++) {
                        JSONObject element = (JSONObject)jsonList.get(i);
                        int getId = element.getInt("id");
                        String getTitle = element.getString("title");
                        adapter.addItem(new RcData(getId,getTitle));
                        Handler handler = new Handler(getMainLooper());
                        handler.post(new Runnable() {
                            @Override
                            public void run() {

                                adapter.notifyDataSetChanged();
                            }
                        });
                    }

                } catch (JSONException e) {
                    e.printStackTrace();
                    Log.e("d", e.toString());
                }

            }
            response.close();
        }
    };

    @Override
    public void setOnData(String title, String director, int year, String synopsis) {
        RequestBody body = new FormBody.Builder()
                .add("title",title)
                .add("director",director)
                .add("year",String.valueOf(year))
                .add("synopsis",synopsis).build();
        Request request = new Request.Builder()

                .url("http://ec2-3-36-106-46.ap-northeast-2.compute.amazonaws.com:3001/movies")
                .post(body)
                .build();
        client.newCall(request).enqueue(updateUserCallBack);
    }
    private Callback updateUserCallBack = new Callback() {
        @Override
        public void onFailure(Call call, IOException e) {

        }

        @Override
        public void onResponse(Call call, Response response) throws IOException {
            String body;
            if (response.body() != null) {
                body = response.body().toString();
                Log.e(TAG, "onResponse: "+body );
            }
        }
    };
}

File: client/Network/kotlin_rest/app/src/main/java/com/example/kotlinrest/MainActivity.kt
package com.example.kotlinrest

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.util.Log
import android.widget.LinearLayout
import android.widget.Toast
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.android.volley.Request
import com.android.volley.Response
import com.android.volley.toolbox.StringRequest
import com.android.volley.toolbox.Volley
import com.example.kotlinrest.databinding.ActivityMainBinding
import com.google.gson.Gson
import kotlin.String as String


class MainActivity : AppCompatActivity() {
    val adapter: JsonAdapter = JsonAdapter()

    val mb by lazy {
        ActivityMainBinding.inflate(layoutInflater)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(mb.root)

        // Instantiate the RequestQueue.
        val queue = Volley.newRequestQueue(this)
        val url = "http://ec2-3-36-106-46.ap-northeast-2.compute.amazonaws.com:3001/movies";

        val layoutManager: RecyclerView.LayoutManager = LinearLayoutManager(this)
        mb.recyclerView.layoutManager = layoutManager

        mb.recyclerView.adapter = adapter


        var getData: String? = ""
        // Request a string response from the provided URL.
        val stringRequest = StringRequest(Request.Method.GET, url,
            { response ->
                // Display the first 500 characters of the response string.
                moviesSuccessGet(response.toString())
                getData = "Response is: ${response.substring(0, 100)}"
                Toast.makeText(applicationContext, getData, Toast.LENGTH_SHORT).show()
            },
            {
                getData = "That didn't work!"
                Toast.makeText(applicationContext, getData, Toast.LENGTH_SHORT).show()
            })
        // Add the request to the RequestQueue.
        queue.add(stringRequest)
        mb.btnTest.setOnClickListener {
            Toast.makeText(applicationContext, "저는 코틀린을 잘하고 싶어요ㅜㅜ", Toast.LENGTH_SHORT).show()
        }
    }

    // url:3001/movies 로 get 요청 보냈을때 Gson
    private fun moviesSuccessGet(response: String) {
        val TAG = "asdfa"
        Log.e(TAG, "moviesSuccessGet: $response")
        val gson: Gson = Gson()
        val jsonMovies = gson.fromJson(response, JsonMovies::class.java)
        val data = jsonMovies.data

        data.forEachIndexed { index, jsonMoviesInfo ->
            adapter.addItem(JsonListData(jsonMoviesInfo.title, jsonMoviesInfo.id))
            adapter.notifyItemInserted(index)

        }


    }
}